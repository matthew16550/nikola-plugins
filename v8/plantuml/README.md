This plugin converts [PlantUML](https://plantuml.com/) files to SVG, PNG or TXT (Ascii Art) files.

There is also a [plantuml_markdown](https://plugins.getnikola.com/#plantuml_markdown) plugin which supports
PlantUML inside Markdown files.

The default configuration will output all `*.puml` files under the `pages` dir as SVG files.

Developed against PlantUML version 1.2020.23.  Probably works with some earlier versions.

# Unicode

The plugin expects PlantUML files to be encoded with UTF-8.

# Known Issues

- It's slow!  Every PlantUML rendering uses a new Java process, on my laptop it takes 4-8 seconds per file.
  I have some ideas to speed this up, and they may be available in future plugin versions.

- Changes to files included via `!include ...` will NOT trigger a rebuild.
  Instead, you could include them via `PLANTUML_ARGS` and those will trigger a rebuild.

- `nikola auto` does not watch dirs in `PLANTUML_FILES` or files included via `PLANTUML_ARGS` / `!include`.
  As a workaround you could put PlantUML files under any dir listed in `POSTS` or `PAGES` because those dirs
  are watched.
  (Use `.iuml` suffix for include files to prevent them matching the `*.puml` wildcard in `PLANTUML_FILES`)

- The line number in PlantUML error messages is one higher than the actual error location.
  Suspect a bug in PlantUML.

- The file name in PlantUML error messages is always `string` rather than the actual problem file.
  PlantUML does this when input is piped via stdin, which as a compromise for simplicity we always do.
