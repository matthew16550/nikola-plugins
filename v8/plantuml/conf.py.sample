#
# PLANTUML_EXEC (list of strings) - The command to run PlantUML
#
# '%site_path%' anywhere in PLANTUML_EXEC will be replaced with the full path to the site dir.
# PlantUML is run in the site dir so often this is not needed.
#
# Examples
# --------
# Run from a JAR file:
#   [ 'java', '-Djava.awt.headless=true', '-jar', 'plantuml.jar' ]
#
# Run in Docker:
#   [ 'docker', 'run', '--interactive', '--rm', '--volume', '%site_path%:/work', 'YOUR_IMAGE',
#     'java', '-Djava.awt.headless=true', '-jar', 'plantuml.jar' ]
#
PLANTUML_EXEC = ['plantuml']

#
# PLANTUML_ARGS (list of strings) - CLI arguments that are sent to PlantUML, see https://plantuml.com/command-line
#
# Examples
# --------
# Use a common style file for all diagrams:
#   [ '-Imy_plantuml_style.iuml' ]
#
# Specify the style in conf.py
#   [ '-chide footbox', '-SShadowing=false' ]
#
PLANTUML_ARGS = []

#
# PLANTUML_FILES contains (wildcard, destination, output_format, args) tuples.
#
# <wildcard> is used to generate a list of source files (in the same way as POSTS and PAGES).
#
# <output_format> is one of 'png', 'svg', 'txt'.
#
# Rendered files will be placed at:
#   output / <destination> / <filename>.<output_format>
#
# As with POSTS and PAGES you can create any directory structure you want and it will be reflected in the output.
#
# <args> is a list of cli arguments that are appended to PLANTUML_ARGS
#
PLANTUML_FILES = (
    ('pages/*.puml', '', 'svg', []),
)

#
# PLANTUML_DEBUG (boolean) - Control verbosity of this plugin
#
PLANTUML_DEBUG = False

#
# PLANTUML_RENDER_ERRORS (boolean) - If True then PlantUML error messages will be rendered in the target files.
#
# You might find this helpful when running `nikola auto`.
#
PLANTUML_RENDER_ERRORS = False
