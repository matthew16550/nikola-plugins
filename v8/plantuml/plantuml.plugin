[Core]
Name = plantuml
Module = plantuml
Tests = test_plantuml

[Nikola]
MinVersion = 8.0.4
PluginCategory = Task

[Documentation]
Author = Matthew Leather
Version = 0.1
Website = https://plugins.getnikola.com/#plantuml
Description = Renders PlantUML files
