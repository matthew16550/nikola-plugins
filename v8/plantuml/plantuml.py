import os
import subprocess
from logging import DEBUG
from math import ceil
from pathlib import Path
from typing import List, Optional
from xml.sax.saxutils import escape

from nikola import Nikola, utils
from nikola.plugin_categories import Task

DEFAULT_ARGS = []

DEFAULT_DEBUG = False

DEFAULT_EXEC = ['plantuml']

DEFAULT_FILES = (
    ('pages/*.puml', '', 'svg', []),
)

DEFAULT_RENDER_ERRORS = True

FORMAT_ARG = {
    'png': '-tpng',
    'svg': '-tsvg',
    'txt': '-tutxt',
}


class PlantUmlTask(Task):
    """Renders PlantUML files"""

    name = 'plantuml'

    def __init__(self):
        super().__init__()
        self.plantuml_manager: Optional[PlantUmlManager] = None

    def set_site(self, site):
        super().set_site(site)
        self.plantuml_manager = PlantUmlManager(site)

    def gen_tasks(self):
        yield self.group_task()

        common_args = self.plantuml_manager.common_args
        filters = self.site.config['FILTERS']
        output_folder = self.site.config['OUTPUT_FOLDER']
        plantuml_files = self.site.config.get('PLANTUML_FILES', DEFAULT_FILES)
        output_path = Path(output_folder)

        # Logic derived from nikola.plugins.misc.scan_posts.ScanPosts.scan()
        for pattern, destination, output_format, args in plantuml_files:
            if output_format not in ['png', 'svg', 'txt']:
                raise Exception("Unsupported PlantUML output format '{}'".format(output_format))
            pattern = Path(pattern)
            root = pattern.parent
            combined_args = common_args + args
            includes = [a[2:] for a in combined_args if a.startswith('-I')]
            kw = {
                'combined_args': combined_args,
                'filters': filters,
                'output_folder': output_folder,
            }

            for src in root.rglob(pattern.name):
                dst = output_path / destination / src.relative_to(root).parent / src.with_suffix('.' + output_format).name
                dst_str = str(dst)
                task = {
                    'basename': self.name,
                    'name': dst_str,
                    'file_dep': includes + [src],
                    'targets': [dst_str],
                    'actions': [(self.render_file, [src, dst, output_format, combined_args])],
                    'uptodate': [utils.config_changed(kw, 'plantuml:' + dst_str)],
                    'clean': True,
                }
                yield utils.apply_filters(task, filters)

    def render_file(self, src: Path, dst: Path, output_format, args: List[str]) -> bool:
        output = self.plantuml_manager.render(src.read_bytes(), output_format, args)
        dst.parent.mkdir(parents=True, exist_ok=True)
        dst.write_bytes(output)
        return True


class PlantUmlManager:
    """
    PlantUmlManager is used by "plantuml" & "plantuml_markdown" plugins
    """

    def __init__(self, site: Nikola):
        self.common_args = site.config.get('PLANTUML_ARGS', DEFAULT_ARGS)
        self._exec = site.config.get('PLANTUML_EXEC', DEFAULT_EXEC)
        self._logger = utils.get_logger('plantuml_manager')
        if site.config.get('PLANTUML_DEBUG', DEFAULT_DEBUG):
            self._logger.level = DEBUG
        self._render_errors = site.config.get('PLANTUML_RENDER_ERRORS', DEFAULT_RENDER_ERRORS)
        self._site = site

    # noinspection PyBroadException
    def render(self, source: bytes, output_format: str, args: List[str]) -> bytes:
        _exec = [e.replace('%site_path%', os.getcwd()) for e in self._exec]
        command = _exec + args + ['-pipe', '-stdrpt', FORMAT_ARG[output_format]]
        self._logger.debug('render() exec: %s\n%s', command, source)
        result = subprocess.run(command, capture_output=True, input=source)

        if result.returncode == 0:
            return result.stdout

        try:
            details = str(result.stderr, encoding='utf8').rstrip()
        except Exception:
            details = str(result.stderr)

        error = "PlantUML error (return code {}): {}".format(result.returncode, details)

        if not self._render_errors:
            raise Exception(error)

        self._logger.exception('Ignoring: %s', error)

        if output_format == 'png':
            return error_to_png(error)
        elif output_format == 'svg':
            return error_to_svg(error)
        else:
            return error.encode('utf8')


def error_to_png(error: str) -> bytes:
    import io
    from PIL import Image, ImageDraw
    size = ImageDraw.Draw(Image.new('RGB', (1, 1))).multiline_textsize(error)
    image = Image.new('RGB', (size[0] + 20, size[1] + 20), color=(0xdc, 0x35, 0x45))
    draw = ImageDraw.Draw(image)
    draw.multiline_text((10, 10), error, fill=(255, 255, 255))
    buf = io.BytesIO()
    image.save(buf, 'PNG')
    return buf.getvalue()


def error_to_svg(error: str) -> bytes:
    # This whole function is kludgy because I don't know SVG very well.  Improvements welcome :-)
    lines = error.split('\n')
    text = '\n'.join('<tspan x="1em" dy="1.2em">{}</tspan>'.format(escape(line)) for line in lines)
    height = ceil(1.2 * (len(lines))) + 2
    width = max(len(line) for line in lines) + 2
    result = '<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n' + \
             '<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="{}em" height="{}em">\n'.format(width, height) + \
             '<rect x="0" y="0" width="{}em" height="{}em" fill="#dc3545"/>\n'.format(width, height) + \
             '<text x="0" y="1em" font-family="monospace" fill="#ffffff">\n' + \
             text + '\n' + \
             '</text>\n' + \
             '</svg>'
    return result.encode('utf8')
