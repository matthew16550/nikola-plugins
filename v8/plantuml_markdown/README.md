Markdown compiler plugin to support [PlantUML](https://plantuml.com/).

Installation
============

Enable the plugin by adding `plantuml_markdown` anywhere BEFORE `fenced_code` in `MARKDOWN_EXTENSIONS` e.g.

    MARKDOWN_EXTENSIONS = ['plantuml_markdown', 'fenced_code', 'codehilite']

The [plantuml](https://plugins.getnikola.com/#plantuml) plugin must also be installed and configured.

Usage
=====

Diagram in the page:

    ``` plantuml
    A -> B : foo
    ```

Listing of PlantUML code:

    ``` { .plantuml listing }
    A -> B : foo
    ```

Diagram with code listing on the right:

    ``` { .plantuml svg+listing }
    A -> B : foo
    ```

Diagram with code listing on the left:

    ``` { .plantuml listing+svg }
    A -> B : foo
    ```

[Fenced Code Block Attributes](https://python-markdown.github.io/extensions/fenced_code_blocks/#attributes)
can also be used.  e.g.

    ``` { .plantuml #my_id svg+listing linenums=y }
    A -> B : foo
    ```

Specify a common prefix for all subsequent PlantUML blocks in the page:

    ``` plantuml_prefix
    skinparam ParticipantBorderColor DeepSkyBlue
    skinparam ParticipantBackgroundColor DodgerBlue
    ```

Known Issues
============

- Changes to include files (`!include ...` or `-I...` in `PLANTUML_ARGS`/ `PLANTUML_MARKDOWN_SVG_ARGS`)
  will NOT trigger a rebuild of associated markdown files. You will need to force a rebuild (`nikola build -a`).
