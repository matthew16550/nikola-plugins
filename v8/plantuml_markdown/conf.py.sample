#
# PLANTUML_MARKDOWN_SVG_ARGS (list of strings) - CLI arguments that are appended to PLANTUML_ARGS (in the 'plantuml' plugin)
#
PLANTUML_MARKDOWN_SVG_ARGS = []
