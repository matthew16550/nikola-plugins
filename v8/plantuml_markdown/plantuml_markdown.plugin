[Core]
Name = plantuml_markdown
Module = plantuml_markdown
Tests = test_plantuml_markdown

[Nikola]
compiler = markdown
MinVersion = 8.0.4
PluginCategory = CompilerExtension

[Documentation]
Author = Matthew Leather
Version = 0.1
Website = https://plugins.getnikola.com/#plantuml_markdown
Description = Renders PlantUML inside Markdown files
