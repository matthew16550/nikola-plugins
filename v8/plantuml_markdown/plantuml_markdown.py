import re
from logging import Logger
from re import Match
from typing import List, Optional

from markdown import Markdown
from markdown.extensions import Extension
from markdown.extensions.attr_list import get_attrs
from markdown.extensions.fenced_code import FencedBlockPreprocessor

from nikola import Nikola
from nikola.plugin_categories import MarkdownExtension

DEFAULT_MARKDOWN_SVG_ARGS = []


class PlantUmlMarkdownExtension(MarkdownExtension, Extension):
    name = 'plantuml_markdown'

    def __init__(self):
        super().__init__()
        self.processor: Optional[PlantUmlMarkdownProcessor] = None

    def extendMarkdown(self, md: Markdown):
        try:
            self.processor = PlantUmlMarkdownProcessor(md, self.site, self.logger)
            md.preprocessors.register(self.processor, 'fenced_code_block_plantuml', 26)
            md.registerExtension(self)
        except TypeError as e:  # Kludge to avoid Extension._extendMarkdown() hiding of TypeErrors
            raise Exception(e)

    def reset(self):
        self.processor.reset()


# Extends FencedBlockPreprocessor but this only handles plantuml blocks,
# all other blocks will be handled elsewhere by the usual "fenced_code" extension.
class PlantUmlMarkdownProcessor(FencedBlockPreprocessor):
    def __init__(self, md: Markdown, site: Nikola, logger: Logger):
        super().__init__(md, {})
        self._logger = logger
        self._plantuml_args: Optional[List[str]] = None
        self._plantuml_manager = None
        self._plantuml_prefix = []
        self._site = site

    def _lazy_init(self):
        if not self._plantuml_manager:
            plantuml_plugin = self._site.plugin_manager.getPluginByName('plantuml', 'Task')
            if not plantuml_plugin:
                raise Exception("'plantuml' plugin is required")
            self.set_plugin_manager(plantuml_plugin.plugin_object.plantuml_manager)

    def set_plugin_manager(self, manager):
        self._plantuml_args = manager.common_args + self._site.config.get('PLANTUML_MARKDOWN_SVG_ARGS', DEFAULT_MARKDOWN_SVG_ARGS)
        self._plantuml_manager = manager

    def reset(self):
        self._plantuml_prefix = []

    def run(self, lines):
        def replace(match):
            lang, _id, classes, config = None, '', [], {}
            if match.group('attrs'):
                _id, classes, config = self.handle_attrs(get_attrs(match.group('attrs')))
                if classes:
                    lang = classes.pop(0)
            else:
                lang = match.group('lang')

            if lang == 'plantuml':
                return self.render_code_block(match, _id, classes, config)
            elif lang == 'plantuml_prefix':
                self._plantuml_prefix = ['-c' + line for line in match.group('code').split('\n')]
                return ''
            else:  # The normal 'fenced_code' plugin will process this later
                return match.group()

        return self.FENCED_BLOCK_RE.sub(replace, '\n'.join(lines)).split('\n')

    def render_code_block(self, match, _id, classes, config):
        def listing():
            return self.render_listing(match)

        def svg():
            return self.render_svg(match)

        div = '<div{}{}>'.format(
            ' id="{}"'.format(_id) if _id else '',
            ' classes="{}"'.format(' '.join(classes)) if classes else '',
        )

        if 'listing+svg' in config:
            html = [div] + listing() + svg() + ['</div>']
        elif 'svg+listing' in config:
            html = [div] + svg() + listing() + ['</div>']
        elif 'listing' in config:
            html = [div] + listing() + ['</div>']
        elif 'svg' in config:
            html = [div] + svg() + ['</div>']
        else:
            html = [div] + svg() + ['</div>']

        return self.md.htmlStash.store('\n'.join(html))

    def render_listing(self, match: Match) -> List[str]:
        lines = [first_line_for_listing_block(match)] + match.group().split('\n')[1:]
        html = super().run(lines)[1]
        return [
            '<div class="plantuml_listing" style="display: inline-block; vertical-align: top">',
            html,
            '</div>'
        ]

    def render_svg(self, match: Match) -> List[str]:
        self._lazy_init()
        args = self._plantuml_args + self._plantuml_prefix
        svg = self._plantuml_manager.render(match.group('code').encode('utf8'), 'svg', args)
        svg = str(svg[svg.index(b'<svg'):], encoding='utf8')  # Remove leading "<?xml version=..."
        return [
            '<div class="plantuml_diagram" style="display: inline-block; vertical-align: top">',
            svg,
            '</div>',
        ]


ID_OR_CLASS_RE = re.compile(' *([.#])([^ =]+) *')


def first_line_for_listing_block(match: Match) -> str:
    def replace(m: Match) -> str:
        return 'anchor_ref={} '.format(m.group(2)) if m.group(1) == '#' else ''

    attrs = match.group('attrs')
    attrs_without_id_or_classes = ID_OR_CLASS_RE.sub(replace, attrs) if attrs else ''

    return ''.join([
        match.group('fence'),
        ' {{ .text {}}}'.format(attrs_without_id_or_classes) if attrs_without_id_or_classes else ' text',
        ' hl_lines="{}"'.format(match.group('hl_lines')) if match.group('hl_lines') else '',
    ])
