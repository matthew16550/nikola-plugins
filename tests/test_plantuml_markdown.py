import os
from textwrap import dedent

import pytest
from approvaltests import verify_with_namer
from approvaltests.pytest.namer import PyTestNamer
from markdown import Markdown
from markdown.extensions.fenced_code import FencedBlockPreprocessor
from pytest import fixture

from v8.plantuml.plantuml import PlantUmlTask
from v8.plantuml_markdown.plantuml_markdown import PlantUmlMarkdownExtension, first_line_for_listing_block


def test_svg(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    ```plantuml
    A -> B : foo
    ```
    """)


def test_with_other_markdown(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    # A Diagram
    ``` plantuml
    A -> B : foo
    ```
    
    # Some Python
    ``` python
    x = 1 + 2
    ```
    """)


def test_listing(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    ```{ .plantuml listing }
    participant A
    participant B
    ```
    """)


def test_line_numbering(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    ```{ .plantuml #baz listing linenums=y linenostart=10 }
    participant A
    participant B
    ```
    """)


def test_svg_and_listing(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    ``` { .plantuml svg+listing }
    participant A
    ```
    """)


def test_listing_and_svg(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    ``` { .plantuml listing+svg }
    participant A
    ```
    """)


def test_prefix(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    ```plantuml_prefix
    skinparam ParticipantBorderColor DeepSkyBlue
    skinparam ParticipantBackgroundColor DodgerBlue
    ```
    
    ```plantuml
    participant A
    ```

    ```plantuml
    participant B
    ```
    """)


def test_error(verify_plantuml_markdown):
    verify_plantuml_markdown("""
    ```plantuml
    this is bad
    ```
    """)


@pytest.mark.parametrize('line, expected', [
    ("```plantuml", "``` text"),
    ("```.plantuml", "``` text"),
    ("```plantuml hl_lines='3 4'", "``` text hl_lines=\"3 4\""),
    ("``` { .plantuml .foo #bar linenums=y hl_lines=\"3 4\" }", "``` { .text anchor_ref=bar linenums=y hl_lines=\"3 4\" }"),
])
def test_first_line_for_listing_block(line, expected):
    match = FencedBlockPreprocessor.FENCED_BLOCK_RE.match(line + "\n```")
    assert match
    assert first_line_for_listing_block(match) == expected


@fixture
def verify_plantuml_markdown(request):
    def f(text):
        text = dedent(text).lstrip()
        site = FakeSite()
        plantuml_task = PlantUmlTask()
        plantuml_task.set_site(site)
        extension = PlantUmlMarkdownExtension()
        extension.set_site(site)
        md = Markdown(extensions=[extension, 'fenced_code', 'codehilite'])
        extension.processor.set_plugin_manager(plantuml_task.plantuml_manager)

        # When
        body = md.convert(text)

        # Then
        html = html_page(body)  # put it in an html page so we can easily see it in a browser
        verify_with_namer(html, PyTestNamer(request, '.html'), encoding='utf8')

    return f


class FakeSite:
    debug = True

    def __init__(self):
        self.config = {
            'PLANTUML_DEBUG': True,
            'PLANTUML_RENDER_ERRORS': True,
            'PLANTUML_MARKDOWN_SVG_ARGS': [
                '-chide footbox',
                '-nometadata',
                '-SDefaultFontName=DejaVu Sans',
                '-SDefaultFontSize=17',
                '-SShadowing=false',
            ],
        }
        if 'PLANTUML_EXEC' in os.environ:
            self.config['PLANTUML_EXEC'] = os.environ['PLANTUML_EXEC'].split()


def html_page(body):
    return dedent('''
        <!doctype html>
        <html lang="en">
          <head>
            <meta charset="utf-8">
            <meta name="viewport" content="width=device-width, initial-scale=1">
            <link href="https://themes.getnikola.com/v8/base/demo/assets/css/rst_base.css" rel="stylesheet" type="text/css">
            <link href="https://themes.getnikola.com/v8/base/demo/assets/css/nikola_rst.css" rel="stylesheet" type="text/css">
            <link href="https://themes.getnikola.com/v8/base/demo/assets/css/code.css" rel="stylesheet" type="text/css">
            <link href="https://themes.getnikola.com/v8/base/demo/assets/css/theme.css" rel="stylesheet" type="text/css">
            <link href="data/plantuml_markdown/local.css" rel="stylesheet" type="text/css">
          </head>
          <body>
            {}
          </body>
        </html>
        ''').lstrip().format(body)
