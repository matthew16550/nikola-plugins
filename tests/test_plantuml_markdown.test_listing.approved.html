<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="https://themes.getnikola.com/v8/base/demo/assets/css/rst_base.css" rel="stylesheet" type="text/css">
    <link href="https://themes.getnikola.com/v8/base/demo/assets/css/nikola_rst.css" rel="stylesheet" type="text/css">
    <link href="https://themes.getnikola.com/v8/base/demo/assets/css/code.css" rel="stylesheet" type="text/css">
    <link href="https://themes.getnikola.com/v8/base/demo/assets/css/theme.css" rel="stylesheet" type="text/css">
    <link href="data/plantuml_markdown/local.css" rel="stylesheet" type="text/css">
  </head>
  <body>
    <div>
<div class="plantuml_listing" style="display: inline-block; vertical-align: top">
<pre class="code literal-block">participant A
participant B
</pre>
</div>
</div>
  </body>
</html>
